File: find_string_grid.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
  bool ok(vector<vector<char>>& g, int i, int j, const string& w) {
    int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};
    int n = w.size();
    for (int k = 0; k < 8; ++k) {
      int ni = i + dx[k] * (n - 1);
      int nj = j + dy[k] * (n - 1);
      if (ni >= 0 && ni < g.size() && nj >= 0 && nj < g[0].size()) {
        int l = 0;
        for (; l < n; ++l) {
          if (w[l] != g[i + dx[k] * l][j + dy[k] * l]) {
            break;
          }
        }
        if (l == n) return true;
      }
    }
    return false;
  }

 public:
  vector<vector<int>> searchWord(vector<vector<char>> grid, string word) {
    // Code here
    vector<vector<int>> ans;

    for (int i = 0; i < grid.size(); ++i) {
      for (int j = 0; j < grid[i].size(); ++j) {
        if (grid[i][j] == word[0] && ok(grid, i, j, word)) {
          ans.push_back(vector<int>{i, j});
        }
      }
    }
    return ans;
  }
};

//{ Driver Code Starts.
int main() {
  int tc;
  cin >> tc;
  while (tc--) {
    int n, m;
    cin >> n >> m;
    vector<vector<char>> grid(n, vector<char>(m, 'x'));
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) cin >> grid[i][j];
    }
    string word;
    cin >> word;
    Solution obj;
    vector<vector<int>> ans = obj.searchWord(grid, word);
    for (auto i : ans) {
      for (auto j : i) cout << j << " ";
      cout << "\n";
    }
    if (ans.size() == 0) {
      cout << "-1\n";
    }
  }
  return 0;
}
// } Driver Code Ends

File: matrix/find_string_grid.py
# User function Template for python3


class Solution:
    def searchWord(self, grid, word):
        # Code here
        n = len(grid)
        m = len(grid[0])
        dx = [-1, -1, -1, 0, 0, 1, 1, 1]
        dy = [-1, 0, 1, -1, 1, -1, 0, 1]
        ans = []

        for i in range(n):
            for j in range(m):
                if grid[i][j] != word[0]:
                    continue
                wn = len(word)
                found = False
                for k in range(8):
                    if (
                        i + (wn - 1) * dx[k] >= 0
                        and i + (wn - 1) * dx[k] < n
                        and j + (wn - 1) * dy[k] >= 0
                        and j + (wn - 1) * dy[k] < m
                    ):
                        idx = 0
                        while idx < wn:
                            ni = i + idx * dx[k]
                            nj = j + idx * dy[k]
                            if word[idx] != grid[ni][nj]:
                                break
                            idx += 1
                        if idx == wn:
                            found = True
                            break

                if found:
                    ans.append((i, j))

        return ans


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    T = int(input())
    for i in range(T):
        n, m = input().split()
        n = int(n)
        m = int(m)
        grid = []
        for _ in range(n):
            cur = input()
            temp = []
            for __ in cur:
                temp.append(__)
            grid.append(temp)
        word = input()
        obj = Solution()
        ans = obj.searchWord(grid, word)
        for _ in ans:
            for __ in _:
                print(__, end=" ")
            print()
        if len(ans) == 0:
            print(-1)

# } Driver Code Ends
